<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1" data-toc-modified-id="Вопрос-1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вопрос 1</a></span></li><li><span><a href="#Вопрос-2" data-toc-modified-id="Вопрос-2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Вопрос 2</a></span></li><li><span><a href="#Вопрос-3" data-toc-modified-id="Вопрос-3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вопрос 3</a></span></li><li><span><a href="#Вопрос-4" data-toc-modified-id="Вопрос-4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вопрос 4</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Исследование надежности заемщиков

В проекте требуется исследовать данные о платёжеспособности клиентов банка, а затем ответить на вопросы: 
    
    1) Есть ли зависимость между количеством детей и возвратом кредита в срок?
 
    2) Есть ли зависимость между семейным положением и возвратом кредита в срок?
 
    3) Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
 
    4) Как разные цели кредита влияют на его возврат в срок?

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Загрузка и обзор данных

Загрузим и посмотрим на весь датасет в целом.

In [1]:
import pandas as pd

In [2]:
try:
    main_data = pd.read_csv('data.csv')
except:
    main_data = pd.read_csv('/datasets/data.csv')
main_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В двух столбцах есть пропуски.
Точно не ясно, есть ли лишние пробелы в конце названий колонок. Посмотрим подробно информацию про колонки таблицы.

In [3]:
main_data.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

С названиями колонок всё хорошо. Посмотрим первые 10 строк данных.

In [4]:
main_data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Для полноты картины посмотрим последние 10 строк.

In [5]:
main_data.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Уже видны аномалии. Их исправление будет производиться далее в одном из шагов.

## Предобработка данных

### Заполнение пропусков

- Пропуски есть в столбцах days_employed и total_income. 
- Строки с пропусками составляют долю в 10% от всей таблицы, поэтому нельзя просто взять и удалить их, так как это повлияет на возможные выводы в проекте.
- Так как в двух столбцах одинаковое количество пропусков, то причина их наличия, скорей всего, технологическая. Но также в строках могут быть указаны клиенты, которые официально безработные (у них нет подтверждённого стажа и дохода). Это как действительно безработные, так и самозанятые, чей доход не фиксируется документами. Но стоит проверить и посмотреть, у кого же имеются пропуски.


In [6]:
main_data[main_data['days_employed'].isna()]

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Предположение про самозанятых и безработных не подтвердилось. Причина пропусков явно технологическая. Как-то при создании таблицы потерялись данные.

Тогда нужно от пропусков избавиться, заполнив их. Оба столбца с пропусками хранят количественные значения. Пропуски следует заполнять медианными значениями, так как во-первых, могут быть аномалии в значениях, а во-вторых, какие-то значения могут сильно отличаться от других. К примеру, сверхвысокий доход у кого-то из клиентов. Расчёт среднего арифметического для заполнения пропусков будет не корректен.

В связи с тем, что столбец days_employed содержит множество аномалий, то пока нельзя производить расчёт медианных значений в этом столбце, они будут искажёнными. Так и вышло на практике. При выполнении проекта я изначально избавился от пропусков, а только потом исправлял аномалии. Но медианные значения для столбца days_employed были некорректными.
Поэтому пришлось немного нарушить шаги в проекте и для столбца days_employed сначала выполнить исправление аномалий.

В столбце days_employed, во-первых, многие значения отрицательные. Во-вторых, у пенсионеров почему-то указан аномальный стаж, который при переводе в годы превышает 100 лет, что невозможно: 

In [7]:
main_data[main_data['income_type']=='пенсионер'].head()

children  days_employed  dob_years education  education_id  \
4          0  340266.072047         53   среднее             1   
12         0            NaN         65   среднее             1   
18         0  400281.136913         53   среднее             1   
24         1  338551.952911         57   среднее             1   
25         0  363548.489348         67   среднее             1   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
12       гражданский брак                 1      M   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
12            NaN                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости

Причина аномальных значений неясна. Возможно стаж был посчитан неверно по какой-либо формуле.

Решение: отрицательные значения в столбце days_employed сделать положительными (взять значение по модулю). 

У пенсионеров можно заменить значения стажа на стаж, вычисленный по формуле: *(dob_years - 22)\*247*. Здесь я делаю допущение, что обычно в 22 года человек заканчивает ВУЗ и идёт работать. 247 - среднее количество рабочих дней в году. Работал ли человек после выхода на пенсию и работает ли сейчас - про это никакой информации нет. В нашей стране многие работают, находясь на пенсии. Поэтому будем считать, что заменяем аномальные значения трудового стажа для работающих по сей день пенсионеров.

In [8]:
main_data['days_employed'] = main_data['days_employed'].abs()
main_data.loc[main_data['income_type']=='пенсионер', 'days_employed'] = (main_data.loc[main_data['income_type']=='пенсионер', 'dob_years'] - 22)*247
main_data[main_data['income_type']=='пенсионер'].head(10)

children  days_employed  dob_years education  education_id  \
4          0         7657.0         53   среднее             1   
12         0        10621.0         65   среднее             1   
18         0         7657.0         53   среднее             1   
24         1         8645.0         57   среднее             1   
25         0        11115.0         67   среднее             1   
29         0        10127.0         63   среднее             1   
30         1         9880.0         62   среднее             1   
35         0        11362.0         68   среднее             1   
50         0        10127.0         63   среднее             1   
55         0         7904.0         54   среднее             1   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
12       гражданский брак                 1      M   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   
29  Не женат / не замужем                 4      F   пенсионер     0   
30        женат / замужем                 0      F   пенсионер     0   
35       гражданский брак                 1      M   пенсионер     0   
50        женат / замужем                 0      F   пенсионер     0   
55       гражданский брак                 1      F   пенсионер     1   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
12            NaN                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости  
29            NaN       строительство жилой недвижимости  
30  171456.067993  операции с коммерческой недвижимостью  
35   77805.677436                  на проведение свадьбы  
50   92342.730612                             автомобили  
55            NaN                        сыграть свадьбу

С виду все значения изменились как и планировалось. 

Теперь можно непосредственно заняться заполнением пропусков. А затем проверим, действительно ли в столбцах не осталось пропусков.

In [9]:
median = main_data['days_employed'].median()
main_data['days_employed'].fillna(value=median, inplace=True)
median = main_data['total_income'].median()
main_data['total_income'].fillna(value=median, inplace=True)
display(main_data['days_employed'].isna().sum())
display(main_data['total_income'].isna().sum())

0

0

Пропуски успешно заполнились. На всякий случай, посмотрим любую строку, в которой раньше были пропуски. И оценим навскидку, верно ли посчитались медианные значения.

In [10]:
main_data.loc[41]

children                                           0
days_employed                            2258.286554
dob_years                                         50
education                                    среднее
education_id                                       1
family_status                        женат / замужем
family_status_id                                   0
gender                                             F
income_type                              госслужащий
debt                                               0
total_income                           145017.937533
purpose             сделка с подержанным автомобилем
Name: 41, dtype: object

С виду всё логично. Стаж около девяти лет и доход примерно 145 тысяч денежных единиц (вероятно, выраженных в рублях).

### Проверка данных на аномалии и исправления

Аномалии в столбце days_emplyed пришлось исправить заранее. Теперь проверим все ли значения адекватные в столбце total_income. Нет ли там огромных заведомо нереальных чисел. Для этого отсортируем таблицу по этому столбцу в порядке возрастания. И посмотрим последние 10 строк.

In [11]:
main_data.sort_values(by='total_income').tail(10)

children  days_employed  dob_years education  education_id  \
11071         1    1851.200013         36    высшее             0   
15268         1   10207.448165         64    высшее             0   
18353         1    3173.282035         41    высшее             0   
18368         1     333.935516         41    ВЫСШЕЕ             0   
17503         0    2285.476482         43   среднее             1   
17178         0    5734.127087         42    высшее             0   
20809         0    4719.273476         61   среднее             1   
9169          1    5248.554336         35   среднее             1   
19606         1    2577.664662         39    высшее             0   
12412         0    1477.438114         44    высшее             0   

               family_status  family_status_id gender income_type  debt  \
11071       гражданский брак                 1      F   сотрудник     0   
15268              в разводе                 3      M   компаньон     0   
18353  Не женат / не замужем                 4      F   компаньон     0   
18368       гражданский брак                 1      M   компаньон     0   
17503        женат / замужем                 0      M   компаньон     0   
17178       гражданский брак                 1      M   компаньон     0   
20809  Не женат / не замужем                 4      F   сотрудник     0   
9169        гражданский брак                 1      M   сотрудник     0   
19606        женат / замужем                 0      M   компаньон     1   
12412        женат / замужем                 0      M   компаньон     0   

       total_income                            purpose  
11071  1.286281e+06  покупка коммерческой недвижимости  
15268  1.350246e+06                              жилье  
18353  1.427934e+06                         автомобиль  
18368  1.551153e+06                            свадьба  
17503  1.597613e+06           операции с недвижимостью  
17178  1.711309e+06                    сыграть свадьбу  
20809  1.715018e+06            покупка жилья для семьи  
9169   1.726276e+06         дополнительное образование  
19606  2.200852e+06         строительство недвижимости  
12412  2.265604e+06                       ремонт жилью

Значения ежемесячного дохода вполне разумные. 

Также я посмотрел первые 10 строк отсортированной таблицы (методом head), там тоже значения корректные, каких-либо отрицательных значений нет (здесь не буду приводить эту команду).

Далее необходимо посмотреть уникальные значения в столбцах с категориальными значениями. Вдруг встретятся некорректные значения. 

In [12]:
display(main_data['children'].value_counts())
display(main_data['dob_years'].value_counts())
display(main_data['education'].value_counts())
display(main_data['education_id'].value_counts())
display(main_data['family_status'].value_counts())
display(main_data['family_status_id'].value_counts())
display(main_data['gender'].value_counts())
display(main_data['income_type'].value_counts())
display(main_data['debt'].value_counts())
display(main_data['purpose'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

0    19784
1     1741
Name: debt, dtype: int64

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

В столбце children присутствуют аномальные значения -1 и 20. Отрицательного количества детей не бывает, а наличие в таблице 76 человек с 20 детьми это странно. Такое количество детей конечно бывает в восточных странах или в Африке, но в данном случае это аномалия. Есть в таблице люди, у которых 5 детей, а потом только сразу 20. Это нереально. Исправим аномалию, заменив значения -1 и 20 на медианные по столбцу. Значения вычисленной медианы надо привести к целому числу, иначе количество детей будет вещественным.

In [13]:
median = int(main_data['children'].median())
main_data.loc[(main_data['children']==-1)|(main_data['children']==20), 'children'] = median
main_data['children'].value_counts()

0    14272
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

Значения исправились верно.

Далее исправим регистр в значениях столбца Education и Family_status

In [14]:
main_data['education'] = main_data['education'].str.lower()
main_data['family_status'] = main_data['family_status'].str.lower()
display(main_data['education'].unique())
display(main_data['family_status'].unique())

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

Регистр значений теперь верный.

В столбце Gender есть одно непонятное значение XNA. Заменим его на преобладающее значение F.

In [15]:
main_data.loc[main_data['gender']=='XNA','gender'] = 'F'
main_data['gender'].unique()

array(['F', 'M'], dtype=object)

 Всё верно. На этом исправление аномалий закончено.

### Изменение типов данных

Изменим вещественный тип данных в столбцах days_employed и total_income на целочисленный.

Использовать здесь команды try...except нет смысла, так как перевод происходит из вещественного числа в целое. А пропусков и строковых значений в столбцах нет.

In [16]:
main_data['days_employed'] = main_data['days_employed'].astype('int64')
main_data['total_income'] = main_data['total_income'].astype('int64')
main_data.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0           7657         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Изменения типов данных прошли корректно. 

### Удаление дубликатов

Посмотрим количество явных дубликатов в таблице.

In [17]:
main_data.duplicated().sum()

71

Откуда появились дубликаты? Сходу неясно. Посмотрим, какие строки являются дубликатами.

In [18]:
main_data[main_data.duplicated()]

children  days_employed  dob_years education  education_id  \
2849          0           2258         41   среднее             1   
3290          0           8892         58   среднее             1   
4182          1           2258         34    высшее             0   
4851          0           9386         60   среднее             1   
5557          0           8892         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0          10374         64   среднее             1   
21032         0           9386         60   среднее             1   
21132         0           2258         47   среднее             1   
21281         1           2258         30    высшее             0   
21415         0           7904         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         145017            покупка жилья для семьи  
3290         145017                    сыграть свадьбу  
4182         145017                            свадьба  
4851         145017                            свадьба  
5557         145017                    сыграть свадьбу  
...             ...                                ...  
20702        145017         дополнительное образование  
21032        145017              заняться образованием  
21132        145017                       ремонт жилью  
21281        145017  покупка коммерческой недвижимости  
21415        145017                  операции с жильем  

[71 rows x 12 columns]

Видно, что дубликатами стали некоторые строки, в которых я ранее устранял пропуски значений. Изначально в таблице эти данные могли быть ошибочно введены с пропуском значений.

Удалим явные дубликаты

In [19]:
main_data = main_data.drop_duplicates().reset_index(drop=True)
main_data.shape

(21454, 12)

In [20]:
main_data.shape

(21454, 12)

Дубликаты успешно удалились. Было в таблице 21525 строк, а теперь 21454.

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

Создадим словарь для уровней образования клиентов. Затем удалим дубликаты в этом словаре.

In [21]:
education_dict = main_data[['education', 'education_id']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
education_dict

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

Теперь создадим словарь для семейного статуса и удалим из него дубликаты.

In [22]:
family_status_dict = main_data[['family_status', 'family_status_id']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Теперь из главной таблицы удалим столбцы education и family_status

In [23]:
main_data = main_data.drop(columns=['education', 'family_status'])
main_data.head(5)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           7657         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

Декомпозиция датафрейма выполнена успешно.

### Категоризация дохода

Создадим функцию для категоризации клиентов по уровню дохода. Затем протестируем работу функции.

In [24]:
def income_group(income):
    if income<=30000:
        return 'E'
    elif income<=50000:
        return 'D'
    elif income<=200000:
        return 'C'
    elif income<=1000000:
        return 'B'
    else:
        return 'A'

display(income_group(0))
display(income_group(1008))
display(income_group(50002))
display(income_group(9999999))

'E'

'E'

'C'

'A'

Теперь применим функцию к столбцу total_income и создадим новый столбец с категорией дохода в таблице.

In [25]:
main_data['total_income_category'] = main_data['total_income'].apply(income_group)
main_data.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           7657         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

Столбец добавился успешно.

### Категоризация целей кредита

Создадим функцию для категоризации целей кредита. 
Посмотрим ещё раз уникальные значения столбца purpose и на основе них можно прописать условия в создаваемой функции. Ещё посчитаем количество уникальных элементов, чтобы проверить, что в функции прописаны все.

In [26]:
display(main_data['purpose'].unique())
print('Длина массива:',main_data['purpose'].nunique())

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Длина массива: 38


In [27]:
def purpose_group(purpose):
    if (purpose=='приобретение автомобиля' or purpose=='на покупку подержанного автомобиля' 
        or purpose=='на покупку своего автомобиля' or purpose=='автомобили'
        or purpose=='сделка с подержанным автомобилем' or purpose=='автомобиль' 
        or purpose=='свой автомобиль' or purpose=='сделка с автомобилем' or purpose=='на покупку автомобиля'):
        return 'операции с автомобилем'
    elif (purpose=='покупка жилья' or purpose=='операции с жильем' or purpose=='покупка жилья для семьи'
        or purpose=='покупка недвижимости' or purpose=='покупка коммерческой недвижимости'
        or purpose=='покупка жилой недвижимости' or purpose=='строительство собственной недвижимости' 
        or purpose=='недвижимость' or purpose=='строительство недвижимости' 
        or purpose=='операции с коммерческой недвижимостью' or purpose=='строительство жилой недвижимости'
        or purpose=='жилье' or purpose=='операции со своей недвижимостью' or purpose=='покупка своего жилья'
        or purpose=='операции с недвижимостью' or purpose=='покупка жилья для сдачи' or purpose=='ремонт жилью'):
        return 'операции с недвижимостью'
    elif purpose=='сыграть свадьбу' or purpose=='на проведение свадьбы' or purpose=='свадьба':
        return 'проведение свадьбы'
    elif (purpose=='дополнительное образование' or purpose=='образование' or purpose=='заняться образованием'
        or purpose=='получение образования' or purpose=='получение дополнительного образования'
        or purpose=='получение высшего образования' or purpose=='профильное образование' 
        or purpose=='высшее образование' or purpose=='заняться высшим образованием'):
        return 'получение образования'
    else:
        return 'не определено'

Теперь можно создавать новый столбец в таблице и применить созданную функцию к столбцу purpose.

In [28]:
main_data['purpose_category'] = main_data['purpose'].apply(purpose_group)
main_data.head(10)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           7657         53             1                 1      F   
5         0            926         27             0                 1      M   
6         0           2879         43             0                 0      F   
7         0            152         50             1                 0      M   
8         2           6929         35             0                 1      F   
9         0           2188         41             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

Столбец добавился успешно, функция отработала правильно.

## Ответы на вопросы

### Вопрос 1
**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Для ответа на вопрос сгруппируем таблицу по количеству детей. Далее следует посчитать общее количество клиентов в группах и количество должников. После этого посчитать процентное соотношение должников к общему числу клиентов в группе. 

In [29]:
total_number = main_data.groupby(by='children')['debt'].count()
debtor_number = main_data.groupby(by='children')['debt'].sum()
debt_ratio = (debtor_number/total_number*100).rename('debt_ratio')
debt_ratio.round(2)

children
0    7.54
1    9.23
2    9.45
3    8.18
4    9.76
5    0.00
Name: debt_ratio, dtype: float64

**Ответ на вопрос 1:** чёткой зависимости между количеством детей и возвратом кредита в срок нет. Да, среди клиентов без детей количество должников меньше,но незначительно (менее чем на 2%). Среди клиентов с самой многодетной семьёй (5 детей) нет должников, но и таких клиентов совсем немного в сравнении с другими. 

### Вопрос 2
**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Для ответа на вопрос сгруппируем таблицу по семейному положению клиентов. Далее посчитаем процентное соотношение должников к общему числу клиентов в группе. Выведем также словарь с семейным статусом.

In [30]:
total_number = main_data.groupby(by='family_status_id')['debt'].count()
debtor_number = main_data.groupby(by='family_status_id')['debt'].sum()
debt_ratio = (debtor_number/total_number*100).rename('debt_ratio')
debt_ratio = debt_ratio.round(2)
pd.DataFrame(debt_ratio).merge(family_status_dict, on='family_status_id')

family_status_id  debt_ratio          family_status
0                 0        7.55        женат / замужем
1                 1        9.35       гражданский брак
2                 2        6.57         вдовец / вдова
3                 3        7.11              в разводе
4                 4        9.75  не женат / не замужем

**Ответ на вопрос 2:** явной зависимости опять не прослеживается. Разница в процентном соотношении должников в группах небольшая. Больше всего должников у тех, кто состоит в гражданском браке или не женат/не замужем. В остальных группах должников немного меньше, это объясняется повышенной ответственностью людей (в браке или после развода/смерти супруга).

### Вопрос 3
**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Для ответа на вопрос сгруппируем таблицу по сформированным заранее категориям дохода клиентов. Далее посчитаем процентное соотношение должников к общему числу клиентов в группе.

In [31]:
total_number = main_data.groupby(by='total_income_category')['debt'].count()
debtor_number = main_data.groupby(by='total_income_category')['debt'].sum()
debt_ratio = (debtor_number/total_number*100).rename('debt_ratio')
debt_ratio.round(2)

total_income_category
A    8.00
B    7.06
C    8.49
D    6.00
E    9.09
Name: debt_ratio, dtype: float64

**Ответ на вопрос 3:** явной зависимости снова нет. Но можно сделать вывод: наибольшее число должников среди клиентов с самым низким доходом (группа E), а наименьшее - среди клиентов с доходами от 30 до 50 тыс. Самые большие доходы у клиентов из группы A на возврате кредита в срок особенно не сказываются.

### Вопрос 4
**Как разные цели кредита влияют на его возврат в срок?**

Выполним те же действия, что и при ответе на предыдущие вопросы. Группировать будем по названиям целей кредита.

In [32]:
total_number = main_data.groupby(by='purpose_category')['debt'].count()
debtor_number = main_data.groupby(by='purpose_category')['debt'].sum()
debt_ratio = (debtor_number/total_number*100).rename('debt_ratio')
debt_ratio.round(2)

purpose_category
операции с автомобилем      9.36
операции с недвижимостью    7.23
получение образования       9.22
проведение свадьбы          8.00
Name: debt_ratio, dtype: float64

**Ответ на вопрос 4:** явной зависимости нет. Но можно сделать вывод, что сильнее всего в худшую сторону на возврат кредита в срок влияют цели "операции с автомобилем" и "получение образования". 

## Общий вывод

Результаты исследования будут учитываться в системе, которая оценивает способность потенциального заёмщика вернуть кредит банку. Так вот, наименьшей способностью вернуть кредит в срок обладают:
- семьи с детьми (от 1 ребёнка и более);
- клиенты, находящие в гражданском браке или в статусе "не женат/не замужем";
- клиенты с ежемесячным доходом до 30 000;
- заёмщики, взявшие кредит для операций с автомобилем или на получение образования.